### PR TITLE
Add ESM to huggingface

### PR DESCRIPTION
**Pull request by liujas000**
# What does this PR do?

Adding ESM-1b to huggingface following the steps in https://huggingface.co/transformers/add_new_model.html

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@sgugger

### CODE REVIEW

File: docs/source/index.rst
Comment (sgugger) [inline docs/source/index.rst:385]: Please also add your model in the main README (in alphabetical order) then run `make fix-copies` on your branch.

File: docs/source/model_doc/esm.rst
Comment (sgugger) [inline docs/source/model_doc/esm.rst:40]:
```suggestion
- ESM was trained with a masked language modeling (MLM) objective.
```

File: docs/source/model_doc/esm.rst
Comment (sgugger) [inline docs/source/model_doc/esm.rst:64]: Why is this one commented out?

File: src/transformers/__init__.py
Comment (sgugger) [inline src/transformers/__init__.py:528]:
```suggestion
```

File: src/transformers/__init__.py
Comment (sgugger) [inline src/transformers/__init__.py:2205]:
```suggestion
        from .models.esm import ESMTokenizerFast
        from .models.fnet import FNetTokenizerFast
```

File: src/transformers/models/esm/configuration_esm.py
Comment (sgugger) [inline src/transformers/models/esm/configuration_esm.py:29]: Can you just copy and paste the code instead of subclassing? We would really like models to be as independent as possible.

File: src/transformers/models/esm/convert_esm.py
Comment (sgugger) [inline src/transformers/models/esm/convert_esm.py:104]: Write an error message here.

File: src/transformers/models/esm/convert_esm.py
Comment (sgugger) [inline src/transformers/models/esm/convert_esm.py:121]: Same here

File: src/transformers/models/esm/convert_esm.py
Comment (sgugger) [inline src/transformers/models/esm/convert_esm.py:127]: Same here

File: src/transformers/models/esm/convert_esm.py
Comment (sgugger) [inline src/transformers/models/esm/convert_esm.py:133]: Same here

File: src/transformers/models/esm/convert_esm.py
Comment (sgugger) [inline src/transformers/models/esm/convert_esm.py:172]: Maybe a little more indication as to what went wrong?

File: src/transformers/models/esm/modeling_esm.py
Comment (sgugger) [inline src/transformers/models/esm/modeling_esm.py:133]: This whole try except over an assert is really weird. Let's maybe just do:
```suggestion
        if pointer.shape != array.shape
            raise ValueError(f"Pointer shape {pointer.shape} and array shape {array.shape} mismatched")
```

File: src/transformers/models/esm/modeling_esm.py
Comment (sgugger) [inline src/transformers/models/esm/modeling_esm.py:213]: This should be cleaned up or uncommented.

File: src/transformers/models/esm/modeling_esm.py
Comment (sgugger) [inline src/transformers/models/esm/modeling_esm.py:507]: We're moving away from asserts in modeling files, so let's do an if xxx and raise the approriate error.
Also, make sure to respect the 119 char limit in the error message.

File: src/transformers/models/esm/modeling_esm.py
Comment (sgugger) [inline src/transformers/models/esm/modeling_esm.py:577]: This will need to updated once #13657 is merged.

File: src/transformers/utils/modeling_auto_mapping.py
Comment (sgugger) [inline src/transformers/utils/modeling_auto_mapping.py:1]: This file has been deleted and should not be in this PR.

File: tests/test_modeling_esm.py
Comment (sgugger) [inline tests/test_modeling_esm.py:491]: To clean up

File: tests/test_modeling_esm.py
Comment (sgugger) [inline tests/test_modeling_esm.py:497]: To clean up

**Review by sgugger — COMMENTED** [review]
Thanks a lot for your PR! I left a few comments but it's already in pretty good shape.

I have a few more general interrogations:
- The tokenizer fast is imported in the main inits and used in the doc, but it does not exist in the files, so you should either add the tokenizer fast or remove any mention of it.
- Do all the model head make sense for this new architecture? From what I understand it's linked to proteins so I don't know if the task-specific heads like multiple choice or question answering really are useful.
- There should be a file to test the tokenizer

**Comment by sgugger** [comment]
It doesn't look like you added your model in the configuration_auto mappings, just the modeling_auto mappings. That's why you get this error.

**Review by patrickvonplaten — APPROVED** [review]
I fixed a couple of things (corrected the config and removed dead code of ESMForQuestionAnswering) - hope that was ok!

PR is good for merge IMO. It would be nice if we could replace the BERT modules with ESM modules in the conversion script and then we should probably also flip the private model repo to public so that the integration test can run :-) 

Should we upload the other checkpoitns as well: https://github.com/facebookresearch/esm#pre-trained-models-

**Review by LysandreJik — APPROVED** [review]
Nice, this looks good to me! Thank you for working on this!

**Comment by patrickvonplaten** [comment]
Hey @liujas000, 

Can we help you in any way to get this PR merged? :-)

**Comment by liujas000** [comment]
@patrickvonplaten sorry for the delay; I will land this week!

**Comment by patrickvonplaten** [comment]
Test failure seems unrelated

File: src/transformers/models/esm/convert_esm.py
Comment (patrickvonplaten) [reply src/transformers/models/esm/convert_esm.py:25]: This would still be nice to change before merging

**Comment by patrickvonplaten** [comment]
Thanks a lot for making this PR more or less mergeable @liujas000 . I think there are just some final comments from @sgugger and @patrickvonplaten to be taken care of and the PR is good to go :-)

**Comment by gianhiltbrunner** [comment]
I'd love to use this model, what still needs to be done to get this merged?

**Comment by patrickvonplaten** [comment]
cc @liujas000 @Rocketknight1

**Comment by Rocketknight1** [comment]
@gianhiltbrunner We're still waiting on an internal review from Facebook from the contributors, I believe! I'll let you know if there's any update.

**Comment by franzigeiger** [comment]
I would also be very happy if this gets merged! Any progress?

**Comment by patrickvonplaten** [comment]
Any updates here?

**Pull request closed without merging**
